File: bin/gasmeter.py
import RPi.GPIO as GPIO
import time
import sqlite3 as lite

GPIO.setmode(GPIO.BOARD)
GPIO.setwarnings(False)

# GPIO define
REED_gas = 21
# define GPIO as INPUT and set it also as PULUPDOWN
GPIO.setup(REED_gas, GPIO.IN, pull_up_down=GPIO.PUD_UP)

# database
database = "/opt/loxberry/data/plugins/gasmeter/gasmeter_database.sqlite3"
db = lite.connect(database)

state_old=1
while True:
    state_actual = GPIO.input(REED_gas)
    # REEDKONTAKT geoeffnet
    if state_actual == 1:
        print "contact open"
        state_old=GPIO.input(REED_gas)
        # REEDKONTAKT geschlossen
    elif state_actual==0:
        print "contact closed"
        if state_old!=state_actual:
            state_old=GPIO.input(REED_gas)
            # Impuls into Database insert
            cursor = db.cursor()
            cursor.execute("""INSERT INTO gasmeter_t (timestamp,counter) VALUES (CURRENT_TIMESTAMP,1)""")
            db.commit()
            cursor.close()
    time.sleep(1)

File: bin/gasmeter.pl
#!/usr/bin/perl

use LoxBerry::System;
use File::HomeDir;

use CGI qw/:standard/;
use Config::Simple qw/-strict/;
use warnings;
use strict;
use LoxBerry::IO;
use LoxBerry::Log;
use Time::HiRes qw ( sleep );

use DBI;
use RPi::Pin;
use RPi::Const qw(:all);


my $log = LoxBerry::Log->new(name => 'Input_handler',);

LOGSTART("Handle input daemon");

my $pcfg = new Config::Simple("$lbpconfigdir/pluginconfig.cfg");
my $database = $pcfg->param("MAIN.DATABASE");
my $samplingRate = $pcfg->param("MAIN.SAMPLINGRATE");
my $gpio = $pcfg->param("INPUT.GPIO");

# set the pin
my $pin = RPI::Pin->new($gpio);
$pin->mode(INPUT);
$pin->pull(PUD_UP);

LOGDEB "Congigured database: $database";
LOGDEB "Configured sampling rate: $samplingRate";
LOGDEB "Configured gpio: $gpio";
LOGDEB "Configured gpio as: " . $pin->mode ;
    
# connect to database
my $dsn = "DBI:SQLite:dbname=$database";
my $user = undef;
my $password = undef;
my $dbh = DBI->connect($dsn,$user,$password, { AutoCommit => 1, RaiseError => 1 }) or die $DBI::errstr;
my $sth = $dbh->prepare("INSERT INTO gasmeter_t (timestamp,counter) VALUES (CURRENT_TIMESTAMP,1)");
	 
LOGDEB "Connected to database: $dat";


my $state_old=0;
my $state_actual=0;

#endless loop
while(1){
    
    $state_actual = $pin->read();
    if ($state_actual) {
	print "contact open";
	$state_old=$pin->read();
    } else { # state_actual=0
	print "contact closed";
	if ($state_old != $state_actual) {
	    $state_old = $pin->read();
	    #write into DB
	    $sth->execute;
	}
    }
    
    #For Loglevel more than Error (e.g. Debug) we set the sleep time to 1sec
    if($log->loglevel() >3){
	sleep(1);
    } else {
	sleep ($samplingRate);
    }
	
}

exit;
END
{
    if ($log) {
        $log->LOGEND;
    }
}

